File: src/eyes/calibration.py
"""Calibration module for computing neutral pose from samples."""

from __future__ import annotations

from dataclasses import dataclass
from typing import NamedTuple


class PoseSample(NamedTuple):
    """A single yaw/roll sample from head pose detection."""

    yaw: float
    roll: float


def compute_median_pose(samples: list[PoseSample]) -> PoseSample:
    """Compute median yaw and roll from a list of pose samples.

    Sorts samples by yaw and returns the middle sample's yaw and roll.
    For even count (2 samples), returns the average of both samples.

    Args:
        samples: List of (yaw, roll) pose samples.

    Returns:
        PoseSample with median yaw and roll.

    Raises:
        ValueError: If samples list is empty.
    """
    if not samples:
        raise ValueError("At least one sample is required to compute median pose")

    sorted_by_yaw = sorted(samples, key=lambda s: s.yaw)
    n = len(sorted_by_yaw)

    if n == 1:
        return sorted_by_yaw[0]

    if n % 2 == 1:
        # Odd count: return the middle element
        mid = n // 2
        return sorted_by_yaw[mid]

    # Even count: average the two middle elements
    mid1 = n // 2 - 1
    mid2 = n // 2
    median_yaw = (sorted_by_yaw[mid1].yaw + sorted_by_yaw[mid2].yaw) / 2
    median_roll = (sorted_by_yaw[mid1].roll + sorted_by_yaw[mid2].roll) / 2
    return PoseSample(yaw=median_yaw, roll=median_roll)


# Epsilon for floating-point countdown comparisons. 10 ticks of 0.1s leaves
# residue around 1e-16; choose a value comfortably above that yet far below
# any meaningful sampling interval.
_COUNTDOWN_EPSILON = 1e-9


@dataclass(frozen=True)
class CalibrationResult:
    """Final outcome of a calibration session."""

    yaw: float
    roll: float
    sample_count: int


class CalibrationSession:
    """Qt-free 5-second calibration session.

    Lifecycle:
        1. Construct with duration (default 5s); session starts inactive.
        2. Call start() to begin the countdown.
        3. Call feed(yaw, roll) from each pose update while active.
        4. Call tick(dt) on each timer tick; countdown decrements.
        5. Once countdown reaches 0, session is finished and inactive.
        6. result() returns the median pose, or None if no samples collected
           or session not finished.

    Independent of Qt — drive it from any timer or test harness.
    """

    def __init__(self, duration_seconds: float = 5.0) -> None:
        self._duration = duration_seconds
        self._countdown = 0.0
        self._samples: list[PoseSample] = []
        self._active = False
        self._finished = False

    @property
    def is_active(self) -> bool:
        return self._active

    @property
    def countdown_seconds(self) -> float:
        return self._countdown

    @property
    def sample_count(self) -> int:
        return len(self._samples)

    def start(self) -> None:
        """Begin a fresh calibration session: clear samples and seed countdown."""
        self._samples = []
        self._countdown = self._duration
        self._active = True
        self._finished = False

    def feed(self, yaw: float, roll: float) -> None:
        """Record a pose sample. No-op when the session is inactive."""
        if not self._active:
            return
        self._samples.append(PoseSample(yaw=yaw, roll=roll))

    def tick(self, dt: float) -> None:
        """Advance the countdown. No-op when the session is inactive.

        When the countdown reaches zero the session deactivates and marks
        itself finished so that result() becomes available.
        """
        if not self._active:
            return
        self._countdown -= dt
        # Tolerate floating-point drift from accumulated 0.1s ticks.
        if self._countdown <= _COUNTDOWN_EPSILON:
            self._countdown = 0.0
            self._active = False
            self._finished = True

    def result(self) -> CalibrationResult | None:
        if not self._finished or not self._samples:
            return None
        median = compute_median_pose(self._samples)
        return CalibrationResult(
            yaw=median.yaw, roll=median.roll, sample_count=len(self._samples)
        )

File: src-tauri/src/domain/calibration.rs
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct PoseSample {
    pub yaw: f64,
    pub pitch: f64,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct CalibrationResult {
    pub yaw: f64,
    pub pitch: f64,
    pub sample_count: usize,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct EmptySamples;

const COUNTDOWN_EPSILON: f64 = 1e-9;

pub fn compute_median_pose(samples: &[PoseSample]) -> Result<PoseSample, EmptySamples> {
    if samples.is_empty() {
        return Err(EmptySamples);
    }

    let mut sorted: Vec<PoseSample> = samples.to_vec();
    sorted.sort_by(|a, b| {
        a.yaw
            .partial_cmp(&b.yaw)
            .unwrap_or(std::cmp::Ordering::Equal)
    });

    let len = sorted.len();
    if len == 1 {
        return Ok(sorted[0]);
    }

    if len % 2 == 1 {
        return Ok(sorted[len / 2]);
    }

    let mid1 = &sorted[len / 2 - 1];
    let mid2 = &sorted[len / 2];
    Ok(PoseSample {
        yaw: (mid1.yaw + mid2.yaw) / 2.0,
        pitch: (mid1.pitch + mid2.pitch) / 2.0,
    })
}

#[derive(Debug, Clone)]
pub struct CalibrationSession {
    duration_seconds: f64,
    countdown: f64,
    samples: Vec<PoseSample>,
    active: bool,
    finished: bool,
}

impl CalibrationSession {
    pub fn new(duration_seconds: f64) -> Self {
        Self {
            duration_seconds,
            countdown: 0.0,
            samples: Vec::new(),
            active: false,
            finished: false,
        }
    }

    pub fn is_active(&self) -> bool {
        self.active
    }

    pub fn countdown_seconds(&self) -> f64 {
        self.countdown
    }

    pub fn sample_count(&self) -> usize {
        self.samples.len()
    }

    pub fn start(&mut self) {
        self.samples.clear();
        self.countdown = self.duration_seconds;
        self.active = true;
        self.finished = false;
    }

    pub fn feed(&mut self, yaw: f64, pitch: f64) {
        if !self.active {
            return;
        }
        self.samples.push(PoseSample { yaw, pitch });
    }

    pub fn tick(&mut self, dt: f64) {
        if !self.active {
            return;
        }
        self.countdown -= dt;
        if self.countdown <= COUNTDOWN_EPSILON {
            self.countdown = 0.0;
            self.active = false;
            self.finished = true;
        }
    }

    pub fn result(&self) -> Option<CalibrationResult> {
        if !self.finished || self.samples.is_empty() {
            return None;
        }
        let median = compute_median_pose(&self.samples).ok()?;
        Some(CalibrationResult {
            yaw: median.yaw,
            pitch: median.pitch,
            sample_count: self.samples.len(),
        })
    }
}

/// 喂入一个校准样本并推进 0.1 秒倒计时。
///
/// 如果会话仍在进行中，返回 `None`；校准结束时返回结果。
/// 对非活跃会话不做任何操作。
pub fn feed_calibration_sample(
    session: &mut CalibrationSession,
    yaw: f64,
    pitch: f64,
) -> Option<CalibrationResult> {
    if !session.is_active() {
        return None;
    }
    session.feed(yaw, pitch);
    session.tick(0.1);
    if session.is_active() {
        None
    } else {
        session.result()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn feed_calibration_sample_returns_none_while_active() {
        let mut session = CalibrationSession::new(1.0);
        session.start();

        // 1 秒倒计时，每 tick 0.1 秒，前 9 次仍在进行中
        for _ in 0..9 {
            assert!(feed_calibration_sample(&mut session, 3.0, 5.0).is_none());
        }
        assert!(session.is_active());
        assert_eq!(session.sample_count(), 9);
    }

    #[test]
    fn feed_calibration_sample_returns_result_when_finished() {
        let mut session = CalibrationSession::new(1.0);
        session.start();

        // 第 10 次 tick 使倒计时归零
        for _ in 0..9 {
            feed_calibration_sample(&mut session, 3.0, 5.0);
        }
        let result = feed_calibration_sample(&mut session, 7.0, 9.0);
        let result = result.expect("校准应已完成");
        assert_eq!(result.sample_count, 10);
        // 所有样本 yaw=3.0 但最后一次 yaw=7.0，中位数为 3.0
        assert!((result.yaw - 3.0).abs() < 1e-9);
    }

    #[test]
    fn feed_calibration_sample_ignores_inactive_session() {
        let mut session = CalibrationSession::new(1.0);
        // 未 start，直接调用
        assert!(feed_calibration_sample(&mut session, 1.0, 2.0).is_none());
        assert_eq!(session.sample_count(), 0);
    }
}
